### PR TITLE
Fix new subscriptions

### PR DESCRIPTION
**Pull request by rohitpaulk**
[Bounty#500](https://assembly.com/coderwall/bounties/500)

I've made the commits small and verbose for reviewing purposes. I'll squash and reword them once this is reviewed.

### CODE REVIEW

File: app/models/teams/account.rb [hunk 1]
```diff
@@ -18,16 +18,13 @@ class Teams::Account < ActiveRecord::Base
   has_many :plans, through: :account_plans
   belongs_to :admin, class_name: 'User'
 
-  validates :team_id, presence: true, uniqueness: true
   validates_presence_of :stripe_card_token
   validates_presence_of :stripe_customer_token
+  validates :team_id, presence: true, uniqueness: true
+  validate :admin_id, :payer_is_team_admin
```
Comment (seuros) [inline app/models/teams/account.rb]: we should remove that.

File: app/models/teams/account.rb [hunk 2]
```diff
@@ -18,16 +18,13 @@ class Teams::Account < ActiveRecord::Base
...
 
   attr_protected :stripe_customer_token, :admin_id
 
-  validate :stripe_customer_token, presence: true
-  validate :stripe_card_token, presence: true
-  validate :admin_id, :payer_is_team_admin
-
   def payer_is_team_admin
     if admin_id.nil? #or !team.admin?(admin)
```
Comment (seuros) [inline app/models/teams/account.rb]: and this

File: app/models/teams/account.rb [hunk 3]
```diff
@@ -18,22 +18,12 @@ class Teams::Account < ActiveRecord::Base
...
   belongs_to :admin, class_name: 'User'
 
-  validates :team_id, presence: true, uniqueness: true
   validates_presence_of :stripe_card_token
   validates_presence_of :stripe_customer_token
+  validates :team_id, presence: true, uniqueness: true
 
   attr_protected :stripe_customer_token, :admin_id
 
-  validate :stripe_customer_token, presence: true
```
Comment (seuros) [inline app/models/teams/account.rb]: what about 
validate :stripe_customer_token, presence: true , if: ->(r){r.state == 'active'}
we need a migration to add add this field :string, default: 'pending'

File: app/models/teams/account.rb [hunk 4]
```diff
@@ -18,22 +18,12 @@ class Teams::Account < ActiveRecord::Base
...
-  validate :stripe_card_token, presence: true
-  validate :admin_id, :payer_is_team_admin
```
Comment (seuros) [inline app/models/teams/account.rb]: we can drop admin_id

File: app/models/teams/account.rb (hunk 3 above)
Comment (rohitpaulk) [inline app/models/teams/account.rb]: Isn't that unnecessary? As far as I understand - Teams who haven't paid, don't have Accounts/Plans. We only create Accounts/Plans for teams who **are** paying customers, so a vanilla `presence` validation makes sense. 

The only way I see a team going to an *inactive* state, is if they decided to cancel their subscription, in which case they should be moved to the free plan. In that case, they'll still have the old `stripe_customer_token`, so our validation still holds.

File: app/models/teams/account.rb (hunk 4 above)
Comment (rohitpaulk) [inline app/models/teams/account.rb]: That is a change that should be made in [#463](https://assembly.com/coderwall/bounties/463). It'll involve modifying the `admin` method too.

Comment (seuros) [inline app/models/teams/account.rb]: :+1:

File: app/models/plan.rb [hunk 5]
```diff
@@ -95,29 +112,8 @@ def unregister_from_stripe
     end
   end
 
-  def price
```
Comment (seuros) [inline app/models/plan.rb]: why did you move those methods ?

Comment (rohitpaulk) [inline app/models/plan.rb]: Just for better organization.

File: app/models/plan.rb [hunk 6]
```diff
@@ -95,29 +112,8 @@ def unregister_from_stripe
...
-  def one_time?
-    self.interval.nil?
-  end
-
-  alias_attribute :has_analytics?, :analytics
-
   #TODO CHANGE with default in rails 4
   def generate_public_id
-    self.public_id = SecureRandom.urlsafe_base64(4).downcase
+    self.public_id = public_id || SecureRandom.urlsafe_base64(4).downcase
```
Comment (seuros) [inline app/models/plan.rb]: what about ``||=`` ?

File: app/models/teams/account.rb [hunk 7]
```diff
@@ -46,10 +36,10 @@ def subscribe_to!(plan, force=false)
   end
 
   def save_with_payment(plan=nil)
-    if valid?
+    if stripe_card_token
       create_customer unless plan.try(:one_time?)
       subscribe_to!(plan) unless plan.nil?
-      team.save!
+      self.save!
```
Comment (seuros) [inline app/models/teams/account.rb]: unneeded self

File: app/models/plan.rb (hunk 6 above)
Comment (rohitpaulk) [inline app/models/plan.rb]: Good idea, fixed.

File: app/models/teams/account.rb (hunk 7 above)
Comment (rohitpaulk) [inline app/models/teams/account.rb]: Fixed.

**Comment by seuros** [comment]
@rohitpaulk could you squash your commits by context ?

**Comment by rohitpaulk** [comment]
Done.
